teps/0161-configmap-as-a-valuesource-in-param-in-taskruns-and-pipelineruns.md

```diff
@@ -0,0 +1,305 @@
+  - [Flexibility](#flexibility)
+  - [Conformance](#conformance)
+  - [User Experience](#user-experience)
+  - [Performance](#performance)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Implementation Plan](#implementation-plan)
+  - [Test Plan](#test-plan)
+  - [Infrastructure Needed](#infrastructure-needed)
+  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This proposal is to be able to reference `ConfigMap` as a value source for a `Param` in `TaskRun` or `PipelineRun` .
+This is to support Kubernetes native options (ConfigMap) as value source along with direct value passed to `TaskRun` and `PipelineRun`.
+
+Note: this proposal is basically picking up upon the closed (unmerged) [proposal](https://github.com/tektoncd/community/pull/868).
+
+## Motivation
+
+`PipelineRun` and `TaskRun` Params support passing direct value during its creation. Two supported options to pass values from ConfigMap are
+- `TEP 0029-step-workspaces`  where ConfigMap is mounted as file inside container.
+- Assigning `Env` from `ConfigMap` reference for each container / step on creation of pipeline.
+
+The unavilability of the value as a parameter can lead to 2 problems:
+- Inability to use the value in the properties that cannot leverage scripts (example: `steps[*].image` in Task)
+
+```yaml
+  steps:
+  - name: build-and-push
+    image: $(params.MAVEN_IMAGE)
+```
+
+- Inability to use a catalog (i.e not defined by the end user) `Pipeline` for the end user's own `PipelineRun` if such a Pipeline has a parameter that the end user can only provide via a ConfigMap.
+
+### Goals
+
+To reference `PipelineRun` or `TaskRun` param value from ConfigMap. 
+
+It has following advantages
+- Dynamically determining the value of the Param at runtime of PipelineRun/TaskRun instead of "compile time" of PipelineRun/TaskRun
+- Promoting reuability of the same PipelineRun/TaskRun definition across different clusters/namespaces by minimizing changes needed (no need to hardcoding the `default` property in `ParamSpec` of a Task/Pipeline or the `value` property in a `Param` in a TaskRun/PipelineRun )
+
+### Non-Goals
+
+This TEP does not plan to add support for `Secret` as a ValueSource given the additional complications about handling parameters as a secret (mainly masking the value in logging and k8s API object definition). The support for Secrets could be addressed in another TEP once the scaffold support for ValueSource in Parameters (this TEP) has been added.
+
+### Use Cases
+
+- Task and Pipeline authors will be able to easily use values from ConfigMaps in properties that do not support a script (example: `steps[*].image` in Task)
+- Catalog users will be able to use configmaps to populate in their PipelineRun/TaskRun values required by the Pipeline/Task
```

Comment:
should we add support for StepActions as well ?

Reply:
This question was resolved in the last meeting. The answer is that this change applies to runtime objects like PipelineRuns and TaskRuns and not configuration objects like StepActions etc.